Session 20 (#15)
javarb/notes: sessions/2018/december/Session_20.md

# SOFTWARE ENGINEERING IN JAVA

## Session 20 (18/12/2018)

- Project creation for API application
- Rebase / merge explanation
- Reseting commits
- See the amount of branches and the differences between the master
- Resolving merge conflicts in API Application

### Notes

#### Project creation

We created a new [project for the API application][1] in order to track the tasks using [kanban method][9].

#### Calculator Frontend

> Roger was explaining to me how to use `rebase` and the difeference with `merge` for the [calculator-frontend repository][2].

Reviewer: ```suggestion
Roger was explaining to me how to use `rebase` and the difference with `merge` for the [calculator-frontend repository][2].
```

> He was also explaining to me why he used an empty branch called `Initial codebase` with an initial commit called `delete-me-after-review` as the beggining of the history. He did that in order to he could raise this [Pull Request (PR) from master against that branch][3] and to compare seeing all changes I did in master, being able to do comments, but with the caveat to couldn't make reviews because was himself who created that branch and raised the PR as he states:

Reviewer: The other way around

```suggestion
He was also explaining to me why he used an initial commit called `Initial codebase` on a branch called `delete-me-after-review` as the begining of the history. He did that in order to he could raise this [Pull Request (PR) from master against that branch][3] and to compare seeing all changes I did in master, being able to do comments, but with the caveat to couldn't make reviews because was himself who created that branch and raised the PR as he states:
```

>git rebase delete-me-after-review
>git push -f
>```

At the end, `delete-me-after-review` branch wasn't neccesary since I applied all changes asked in that branch's PR in another branch I had created called `frontend-bs4`, so I [merged that branch into master][4] (master was only different to the empty branch for 1 commit - the merge one). And the empty branch could be deleted:

```bash
[jaar@port-staff calculator-frontend]$ git status
On branch frontend-bs4
Your branch is up to date with 'origin/frontend-bs4'.
...
        modified:   index.html
        modified:   public/scripts/calculator.js
        modified:   public/stylesheets/calculator.css

[jaar@port-staff calculator-frontend]$ git branch -a
* frontend-bs4
  master
  remotes/origin/HEAD -> origin/master
  remotes/origin/delete-me-after-review
  remotes/origin/frontend-bs4
  remotes/origin/master

[jaar@port-staff calculator-frontend]$ git add . ; git commit -m 'Apply sugesstions in PR #2 #3'; git push
[frontend-bs4 182e403] Apply sugesstions in PR #2 #3
 3 files changed, 10 insertions(+), 56 deletions(-)
...
To github.com:javarb/calculator-frontend.git
   081423d..182e403  frontend-bs4 -> frontend-bs4
```

#### API

> Several operation were made in order to apply the last changes in the most secuential and ordered way into Git history

Reviewer: ```suggestion
Several operation were made in order to apply the last changes in the most sequential and ordered way into Git history
```

Counting objects: 100% (31/31), done.
Delta compression using up to 4 threads
Compressing objects: 100% (8/8), done.
Writing objects: 100% (17/17), 2.53 KiB | 863.00 KiB/s, done.
Total 17 (delta 3), reused 0 (delta 0)
remote: Resolving deltas: 100% (3/3), completed with 3 local objects.
To github.com:javarb/api.git
   3e3245d..c6fe812  http-integration-tests -> http-integration-tests
```

Fetch to sync local master with remote master and tree of remote braches locally:

```bash
[jaar@port-staff api]$ git fetch
remote: Enumerating objects: 2, done.
remote: Counting objects: 100% (2/2), done.
remote: Compressing objects: 100% (2/2), done.
remote: Total 2 (delta 0), reused 0 (delta 0), pack-reused 0
Unpacking objects: 100% (2/2), done.
From github.com:javarb/api
   f77c479..37c4a2d  master     -> origin/master

[jaar@port-staff api]$ git branch -a
  api-basics
  api-calculator
  api-calculator-fix-bug
  bugsnag-integration
* http-integration-tests
  master
  remotes/origin/api-basics
  remotes/origin/api-calculator
  remotes/origin/api-calculator-fix-bug
  remotes/origin/bugsnag-integration
  remotes/origin/http-integration-tests
  remotes/origin/master
```

> And we add the `-p` flag to sync deleting to avoid to see locally, remote branchs that doesn't existed in really remotely:

Reviewer: Could you please note down that this `-p` flag has nothing to do with the `git add -p` flag, and that this one stands for `prune`, whereas the `git add -p` one stands for `partial`.
